common/rpc/direct_peer_chooser.go

```diff
@@ -0,0 +1,112 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package rpc
+
+import (
+	"context"
+	"sync"
+
+	"go.uber.org/yarpc/api/peer"
+	"go.uber.org/yarpc/api/transport"
+	"go.uber.org/yarpc/peer/direct"
+	"go.uber.org/yarpc/yarpcerrors"
+
+	"github.com/uber/cadence/common/dynamicconfig"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/membership"
+)
+
+// directPeerChooser is a peer.Chooser that chooses a peer based on the shard key.
+// Peers are managed by the peerList and peers are reused across multiple requests.
+type directPeerChooser struct {
+	serviceName          string
+	logger               log.Logger
+	t                    peer.Transport
+	enableConnRetainMode dynamicconfig.BoolPropertyFn
+	legacyChooser        peer.Chooser
+	mu                   sync.RWMutex
+}
+
+func newDirectChooser(serviceName string, t peer.Transport, logger log.Logger, enableConnRetainMode dynamicconfig.BoolPropertyFn) *directPeerChooser {
+	return &directPeerChooser{
+		serviceName:          serviceName,
+		logger:               logger,
+		t:                    t,
+		enableConnRetainMode: enableConnRetainMode,
+	}
+}
+
+// Start statisfies the peer.Chooser interface.
+func (*directPeerChooser) Start() error {
+	return nil // no-op
+}
+
+// Stop statisfies the peer.Chooser interface.
+func (*directPeerChooser) Stop() error {
+	return nil // no-op
+}
+
+// IsRunning statisfies the peer.Chooser interface.
+func (*directPeerChooser) IsRunning() bool {
+	return true // no-op
+}
+
+// Choose returns an existing peer for the shard key.
+func (g *directPeerChooser) Choose(ctx context.Context, req *transport.Request) (peer peer.Peer, onFinish func(error), err error) {
+	if g.enableConnRetainMode != nil && !g.enableConnRetainMode() {
+		return g.chooseFromLegacyDirectPeerChooser(ctx, req)
+	}
+
+	if req.ShardKey == "" {
+		return nil, nil, yarpcerrors.InvalidArgumentErrorf("chooser requires ShardKey to be non-empty")
+	}
+
+	// TODO: implement connection retain mode
+	return nil, nil, yarpcerrors.UnimplementedErrorf("direct peer chooser conn retain mode unimplemented")
+}
+
+func (g *directPeerChooser) UpdatePeers(members []membership.HostInfo) {
+	// TODO: implement
+	g.logger.Debug("direct peer chooser got a membership update", tag.Counter(len(members)))
+}
+
+func (g *directPeerChooser) chooseFromLegacyDirectPeerChooser(ctx context.Context, req *transport.Request) (peer.Peer, func(error), error) {
+	g.mu.RLock()
+
+	if g.legacyChooser != nil {
+		g.mu.RUnlock()
+		return g.legacyChooser.Choose(ctx, req)
+	}
+
+	g.mu.RUnlock()
+
+	var err error
+	g.mu.Lock()
+	g.legacyChooser, err = direct.New(direct.Configuration{}, g.t)
+	g.mu.Unlock()
```

Comment:
does `direct.New` try to form a connection synchronously, or is this basically just a struct allocation?
(I imagine this code will change when it becomes real, just bringing it up for visibility)

Reply:
`direct.New` just creates Chooser object. `chooser.Choose` is the one doing conn open/close operations based on request's ShardKey.